lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +533,88 @@ def internal_check_future_reserved_word(o, name)
     end
   end
 
+  def internal_check_file_scope(o, name, file)
+    return if file.nil?
+
+    file_scope = @file_scope_hash[file]
+    if file_scope.nil?
+      return if @file_scope_hash.key?(file)
+      file_scope = @file_scope_hash[file] = generate_file_scope file
+      return if file_scope.nil?
+    end
+
+    if check_scope(name, file_scope)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+    #Return to direct call after benchmarking check method
+    #if name.start_with? file_scope
+    #if name !~ file_scope
+    #  acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    #end
+  end
+
+  def create_scope_match(names)
+    qualified_name = names.join("::")
+
+    #qualified_name
+    Regexp.new "^#{qualified_name}", Regexp::IGNORECASE
+  end
+
+  def check_scope(name, file_scope)
+    #!name.start_with?(file_scope)
+    name !~ file_scope
+  end
+
+  def generate_file_scope(file)
+    path = Pathname.new(file)
+
+    return nil if path.extname != ".pp"
+
+    return nil if initial_manifest(file)
+
+    path = Pathname.new(file).each_filename.to_a
+
+    definition_dir_index = find_module_definition_dir path
+
+    # How can we get this result?
+    # If it is not an initial manifest, it must come from a module,
+    # and from the manifests dir there.  This may never get used...
+    return NEVER_MATCH if definition_dir_index.nil?
+
+    before_definition_dir = definition_dir_index == 0 ? 0 : definition_dir_index - 1
+    after_definition_dir = definition_dir_index + 1
+    names = path[after_definition_dir .. -2] # Directories inside module
+    names.unshift(path[before_definition_dir]) # Name of the module itself
+    # Do not include name of module init file at top level of module
+    if !(path.length == (after_definition_dir+1) && path[-1] == 'init.pp')
+      names.push(path[-1][0 .. -4]) # Remove .pp from filename
+    end
+
+    create_scope_match names
+  end
+
+  def initial_manifest(file)
+    manifest_setting = Puppet[:manifest]
+    full_path = Pathname.new(file).expand_path.to_s
+
+    #Does this ever happen outside of tests?
+    if manifest_setting.nil?
+      return full_path.end_with?('site.pp')
+    end
+
+    full_path == manifest_setting || full_path.start_with?(manifest_setting)
+  end
+
+  # Instead of 'manifests' our file could come from 'functions' or 'types'
+  def find_module_definition_dir(path)
+    reverse_index = path.reverse_each.find_index do |dir|
+      dir == 'manifests' || dir == 'functions' || dir == 'types'
+    end
+    return reverse_index.nil? ? nil : (path.length - 1) - reverse_index
+  end
+
+  # The best option from https://stackoverflow.com/questions/1723182/a-regex-that-will-never-be-matched-by-anything
+  NEVER_MATCH = /^\b$/
```

Comment:
Since a switch to start_with? is recommended, this needs to be the string `"\b".freeze` - i.e. never match a leading Backspace.

Reply:
Also, slightly better to move this to a line before it is used (for example line 567 in the current version).